frontend/src/metabase-lib/filter.ts

```diff
@@ -374,25 +375,136 @@ export function relativeDateFilterParts(
 }
 
 export function excludeDateFilterClause(
-  query: Query,
-  stageIndex: number,
-  { operator, column, values, bucket: bucketName }: ExcludeDateFilterParts,
+  filterParts: ExcludeDateFilterParts,
 ): ExpressionClause {
-  if (!bucketName) {
-    const columnWithoutBucket = withTemporalBucket(column, null);
-    return expressionClause(operator, [columnWithoutBucket]);
-  }
+  // As there are still filter expressions with inline bucketed columns,
+  // we need to remove temporal bucketing manually. When all filter expressions
+  // are migrated to use MBQL functions instead of bucketing we can drop this.
+  // See (metabase#50238) for details.
+  const column = withTemporalBucket(filterParts.column, null);
+
+  return match(filterParts)
+    .with({ bucket: P.nullish }, ({ operator }) =>
+      expressionClause(operator, [column]),
+    )
+    .with({ bucket: "hour-of-day" }, ({ operator, values }) =>
+      expressionClause(operator, [
+        expressionClause("get-hour", [column]),
+        ...values,
+      ]),
+    )
+    .with({ bucket: "day-of-week" }, ({ operator, values }) =>
+      expressionClause(operator, [
+        expressionClause("get-day-of-week", [column, "iso"]),
+        ...values,
+      ]),
+    )
+    .with({ bucket: "month-of-year" }, ({ operator, values }) =>
+      expressionClause(operator, [
+        expressionClause("get-month", [column]),
+        ...values,
+      ]),
+    )
+    .with({ bucket: "quarter-of-year" }, ({ operator, values }) =>
+      expressionClause(operator, [
+        expressionClause("get-quarter", [column]),
+        ...values,
+      ]),
+    )
+    .exhaustive();
+}
 
-  const bucket = findTemporalBucket(query, stageIndex, column, bucketName);
-  const columnWithBucket = withTemporalBucket(column, bucket ?? null);
-  const serializedValues = values.map(value =>
-    serializeExcludeDatePart(value, bucketName),
+export function excludeDateFilterParts(
+  query: Query,
+  stageIndex: number,
+  filterClause: FilterClause,
+): ExcludeDateFilterParts | null {
+  return (
+    expressionExcludeDateFilterParts(query, stageIndex, filterClause) ??
+    legacyTemporalBucketExcludeDateFilterParts(query, stageIndex, filterClause)
   );
+}
 
-  return expressionClause(operator, [columnWithBucket, ...serializedValues]);
+function isDateOrDateTimeColumnMetadata(arg: unknown): arg is ColumnMetadata {
+  return isColumnMetadata(arg) && isDateOrDateTime(arg);
 }
 
-export function excludeDateFilterParts(
+function expressionExcludeDateFilterParts(
+  query: Query,
+  stageIndex: number,
+  filterClause: FilterClause,
+): ExcludeDateFilterParts | null {
+  const filterParts = expressionParts(query, stageIndex, filterClause);
+  return match(filterParts)
```

Comment:
This is super powerful. Imagine writing this code by hand instead of relying on pattern matching. I'll migrate other filter clauses to `ts-pattern` if this gets merged.